python/paddle/fluid/incubate/fleet/collective/__init__.py

```diff
@@ -0,0 +1,165 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+
+import sys
+import logging
+
+import paddle.fluid as fluid
+import paddle.fluid.io as io
+from paddle.fluid.optimizer import Optimizer
+import paddle.fluid.transpiler.distribute_transpiler as dist_transpiler
+
+from ..base.fleet_base import Fleet
+from ..base.fleet_base import Mode
+from ..base.fleet_base import DistributedOptimizer
+
+
+class Collective(Fleet):
+    def __init__(self):
+        super(Collective, self).__init__(Mode.COLLECTIVE)
+        self.local_ip_ = 0
+
+    def init(self, role_maker=None):
+        super(Collective, self).init(role_maker)
+        self._role_maker._generate_role()
+
+    def init_worker(self, executor):
+        logging.warn(
+            "You should not call 'init_worker' method for collective mode.")
+
+    def run_worker(self, executor, main_program=None):
+        logging.warn(
+            "You should not call 'run_worker' method for collective mode.")
+
+    def init_server(self, executor, model_dir=None):
+        logging.warn(
+            "You should not call 'init_server' method for collective mode.")
+
+    def run_server(self, executor):
+        logging.warn(
+            "You should not call 'run_server' method for collective mode.")
+
+    def stop_worker(self):
+        logging.warn(
+            "You should not call 'stop_worker' method for collective mode.")
+
+    def stop(self, executor):
+        """
+        stop(): will be called after a user finishes his/her training task.
+        """
+        self.role_maker_.barrier_all()
+        self.role_maker_.finalize()
+
+    def distributed_optimizer(self, optimizer, strategy=None):
+        if not isinstance(optimizer, Optimizer):
+            raise ValueError("optimizer must be an instance of Optimizer")
+        self.optimizer = CollectiveOptimizer(optimizer, strategy)
+        return self.optimizer
+
+    def save_inference_model(self,
+                             executor,
+                             dirname,
+                             feeded_var_names=None,
+                             target_vars=None,
+                             main_program=None,
+                             export_for_deployment=True):
+        io.save_inference_model(dirname, feeded_var_names, target_vars,
+                                executor, main_program, None, None,
+                                export_for_deployment)
+
+    def save_persistables(self, executor, dirname, main_program=None):
+        io.save_persistables(executor, dirname, main_program, None)
+
+    def _get_worker_endpoints(self):
+        return self.worker_endpoints
+
+    def _get_current_id(self):
+        return self.current_id
```

Comment:
Should we expose the two methods to users (get_worker_endpoints and get_current_id).
To use paddle.fluid.compiler for distributed training, we have to set build_strategy.num_trainers and build_strategy.trainer_id. So, if we expose the above two methods to users, we can simplify the programming for users.